File: matlab-client/test/testSubmittedProblem.m
% Copyright © 2019 D-Wave Systems Inc.
% The software is licensed to authorized users only under the applicable license agreement.  See License.txt.

function test_suite = testSubmittedProblem %#ok<STOUT,*DEFNU>
initTestSuite
end



function testRemoteStatus
sp.type = 'remote';
sp.handle = 'the handle';
status = sapiAsyncStatus(sp);
assertEqual(status, sapiremote_status(sp))
end


function testRemoteStatusTimeReceived
sp.type = 'remote';
sp.handle.test_status.submitted_on = '1:23';
status = sapiAsyncStatus(sp);
assertEqual(status, struct('time_received', '1:23'))

sp.handle.test_status.solved_on = 'later';
status = sapiAsyncStatus(sp);
assertEqual(status, ...
    struct('time_received', '1:23', 'time_solved', 'later'))

sp.handle.test_status = rmfield(sp.handle.test_status, 'submitted_on');
status = sapiAsyncStatus(sp);
assertEqual(status, struct('time_solved', 'later'))
end


function testRemoteRetry
sapiremote_retry('test:reset');
sp.type = 'remote';
sp.handle = 'the handle';
sapiAsyncRetry(sp);
srhist = sapiremote_retry('test:reset');
assertEqual(srhist, {{sp.handle}})
end


function testLocalStatus
sp.type = 'local';
status = sapiAsyncStatus(sp);
assertEqual(status, struct('state', 'DONE'))
end


function testLocalRetry
sapiremote_retry('test:reset');
sp.type = 'local';
sp.handle = 'the handle';
sapiAsyncRetry(sp);
assertTrue(isempty(sapiremote_retry('test:reset')))
end

File: python-client/test/test_submitted_problem.py
# Copyright © 2019 D-Wave Systems Inc.
# The software is licensed to authorized users only under the applicable license agreement.  See License.txt.

from flexmock import flexmock
from dwave_sapi2.local import _LocalSubmittedProblem
from dwave_sapi2.remote import _RemoteSubmittedProblem


def test_remote_retry():
    rsp = flexmock()
    rsp.should_receive('retry')
    sp = _RemoteSubmittedProblem(rsp)
    sp.retry()


def test_remote_status():
    rsp = flexmock()
    rsp.should_receive('status').and_return('the status')
    sp = _RemoteSubmittedProblem(rsp)
    assert sp.status() == 'the status'


def test_remote_status_times():
    rsp = flexmock()
    (rsp.should_receive('status')
        .and_return({'submitted_on': '1:23'})
        .and_return({'solved_on': '4:56'})
        .and_return({'submitted_on': 'earlier', 'solved_on': 'now'}))
    sp = _RemoteSubmittedProblem(rsp)

    assert sp.status() == {'time_received': '1:23'}
    assert sp.status() == {'time_solved': '4:56'}
    assert sp.status() == {'time_received': 'earlier', 'time_solved': 'now'}


def test_local_retry():
    sp = _LocalSubmittedProblem(None, None, None, None)
    sp.retry()


def test_local_status():
    sp = _LocalSubmittedProblem(None, None, None, None)
    assert sp.status() == {'state': 'DONE'}
